### ISSUE TITLE
Pre-Processor without Conditional Compilation of #ifdef 

### ISSUE THREAD

**Issue by VenoVeno** [eclipse-cdt/cdt#353]
Hi..
When the following C source is parsed with the Eclipse CDT Parser [link](https://github.com/VenoVeno/CDT-Failed-Demo).

### Source:
```
#include <stdio.h>

int main()
{
#ifdef GOOD
    printf("Good");
#else
    printf("Bad");
#endif
}
```

### Output:
```
- CASTTranslationUnit
  - CASTFunctionDefinition
    - CASTSimpleDeclSpecifier
    - CASTFunctionDeclarator
      - CASTName
    - CASTCompoundStatement
      - CASTExpressionStatement
        - CASTFunctionCallExpression
          - CASTIdExpression
            - CASTName
          - CASTLiteralExpression
```

Here GOOD macro is not defined anywhere in source, #else part will be parsed and included in TU. Is there any way to include both the #ifdef part and #else part, Without doing conditional compilation at Pre-Processing stage. (More like and if else statement) 

Any solution is appreciated. Thanks.

**Comment by jonahgraham** [comment]
I don't think there is any solution to this in CDT at the moment, but I am not sure so will leave this issue open for a bit. When in the Editor context there is some additional functionality such as syntax highlighting and open declaration, but I think those don't use the AST.

**Comment by VenoVeno** [comment]
Hi @jonahgraham. Thanks for responding. However, the AST should be a tree for all possibles right? Is there any possible workaround to skip the conditional checks anywhere in the Preprocessing stage?.

**Comment by jonahgraham** [comment]
> However, the AST should be a tree for all possibles right

No, I don't think it should be. The AST is for the possibility selected based on current pre-processor decisions.

In your example when GOOD isn't defined the contents of that macro still parse sucessfully, but there are many examples when if you visited the not taken path it would lead to having syntax errors.

**Comment by VenoVeno** [comment]
> the contents of that macro still parse sucessfully

Is there a way to include that parsed content to the AST's Translation Unit?

> but there are many examples when if you visited the not taken path it would lead to having syntax errors.

Yeah, i understand that part.

**Comment by VenoVeno** [comment]
Any update on this?

**Comment by jonahgraham** [comment]
> Any update on this?

I don't think there is anything else to add. But I can try to continue the conversation to make sure we are on the same page.

> > the contents of that macro still parse sucessfully
> 
> Is there a way to include that parsed content to the AST's Translation Unit?
>
> > but there are many examples when if you visited the not taken path it would lead to having syntax errors.
> 
> Yeah, i understand that part.

But this is the only part -  hen I said that contents could be parsed successfully, I was talking in a theoretical way, i.e. the syntax within your example excluded block could be included syntactically and semantically with the surrounding code. Because in general it isn't valid and is often syntactically not parseable AFAIK it isn't parsed. So the short answer to "Is there a way to include that parsed content to the AST's Translation Unit?" is no (or at least not in anyway I know) because it *isn't* parsed.

If you want to make a feature request, with the cooresponding PR, then please go ahead - although I am not sure it is a worthwhile eneavour. Otherwise I don't think there is anything left to do on this issue.

It may help the community understand how to help you better if you can explain what the problem you are trying to solve is.

**Comment by VenoVeno** [comment]
> I was talking in a theoretical way, i.e. the syntax within your example excluded block could be included syntactically and semantically with the surrounding code. Because in general it isn't valid and is often syntactically not parseable AFAIK it isn't parsed.

Means, if the branch includes the wrong syntax (Syntactically error code), taking both the branch will result in syntax error. Correct me if i'm wrong about this statement.

> If you want to make a feature request, with the cooresponding PR, then please go ahead - although I am not sure it is a worthwhile eneavour. Otherwise I don't think there is anything left to do on this issue.

I will give it a try..

> It may help the community understand how to help you better if you can explain what the problem you are trying to solve is

My problem here is, the source i'm trying to parse support multiple OS version, Windows, DOS, MAC and Linux. Those different one follows different syntax. The switch b/w the OS branches based on macro definition.

```
#ifdef MSDOS
	sfx = func(bak);
#else
	sfx = value + strlen(value);
#endif
```

In the above case, if the MSDOS is defined it takes the first `sfx = func(bak);` else `sfx = value + strlen(value);`. But i use this AST to do Static code analysis. I need to check for all possible branches like an if else statement.

**Comment by VenoVeno** [comment]
Any update on the above one?

**Comment by jonahgraham** [comment]
No update, nothing has really changed. We are indeed talking about the same thing.

**Comment by VenoVeno** [comment]
Hi @jonahgraham.

Yeah we were discussing about the same thing. I have been using the Eclipse CDT AST to collect data and do SAST analysis on top of it. But as Eclipse cdt doesn't have full support for C++ (https://github.com/eclipse-cdt/cdt/issues/432). 

If everything meet the basic requirement. Will try to add the missing features that CDT lacks.

Thanks :smile:
